test_runner/regress/test_remote_storage.py

```diff
@@ -777,6 +777,103 @@ def create_in_background():
         create_thread.join()
 
 
+# Regression test for a race condition where files are compactified before the upload,
+# resulting in the uploading complaining about the file not being found
+# https://github.com/neondatabase/neon/issues/4526
+@pytest.mark.parametrize("remote_storage_kind", [RemoteStorageKind.LOCAL_FS])
+def test_compaction_delete_before_upload(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_compaction_delete_before_upload",
+    )
+
+    env = neon_env_builder.init_start()
+
+    # create tenant with config that will determinstically allow
+    # compaction and disables gc
+    tenant_id, timeline_id = env.neon_cli.create_tenant(
+        conf={
+            # Set a small compaction threshold
+            "compaction_threshold": "3",
+            # Disable GC
+            "gc_period": "0s",
+            # disable PITR
+            "pitr_interval": "0s",
+        }
+    )
+
+    client = env.pageserver.http_client()
+
+    endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
```

Comment:
In general I would always prefer 
```suggestion
    with env.endpoints.create_start("main", tenant_id=tenant_id) as endpoint:
        # everything up to:
    
    q: queue.Queue[...] = ...
```

Because leaving the compute running opens up a chance of it producing more WAL, which has been a cause of a flakyness in some tests. However this test is not sensitive to that, because we are interested in the single checkpoint + compaction pair.

But this is something to keep in mind. I'll suggest something else to make this a bit smaller one.